File: console/src/services/revenue.ts
/**
 * Revenue Service - Multi-stream revenue generation and income tracking
 */

import axios from 'axios';
import { RevenueStats, RevenueStream, UserRewardParams } from '../types';

export class RevenueService {
  private apiEndpoint: string;

  constructor(apiEndpoint: string = 'http://localhost:5000') {
    this.apiEndpoint = apiEndpoint;
  }

  /**
   * Get comprehensive revenue statistics
   */
  async getRevenueStats(): Promise<RevenueStats> {
    try {
      const response = await axios.get(`${this.apiEndpoint}/revenue/stats`);
      return response.data;
    } catch (error) {
      // Return stub data if API unavailable
      return this.getStubRevenueStats();
    }
  }

  /**
   * Calculate user rewards with multipliers
   */
  async calculateUserRewards(params: UserRewardParams): Promise<string> {
    try {
      const response = await axios.post(`${this.apiEndpoint}/revenue/calculate`, params);
      return response.data.calculated_reward;
    } catch (error) {
      // Local calculation if API unavailable
      return this.calculateRewardsLocal(params);
    }
  }

  /**
   * Local reward calculation (matches Python implementation)
   */
  private calculateRewardsLocal(params: UserRewardParams): string {
    const userStake = parseFloat(params.userStake);
    const totalStaked = parseFloat(params.totalStaked);

    if (totalStaked === 0 || userStake === 0) {
      return '0';
    }

    // Base share proportional to stake
    let baseShare = userStake / totalStaked;
    let multiplier = 1.0;

    // Long-term holder bonus
    if (params.holdingMonths >= 24) {
      multiplier *= 1.5;
    } else if (params.holdingMonths >= 12) {
      multiplier *= 1.25;
    } else if (params.holdingMonths >= 6) {
      multiplier *= 1.1;
    }

    // Active forger bonus
    if (params.forgeCount >= 100) {
      multiplier *= 1.3;
    } else if (params.forgeCount >= 50) {
      multiplier *= 1.15;
    } else if (params.forgeCount >= 10) {
      multiplier *= 1.05;
    }

    // Liquidity provider bonus
    if (params.isLp) {
      multiplier *= 1.2;
    }

    const weightedShare = baseShare * multiplier;
    return weightedShare.toFixed(8);
  }

  /**
   * Get performance metrics for a specific revenue stream
   */
  async getStreamPerformance(streamName: string): Promise<any> {
    try {
      const response = await axios.get(
        `${this.apiEndpoint}/revenue/stream/${streamName}`
      );
      return response.data;
    } catch (error) {
      return { error: 'Stream not found or API unavailable' };
    }
  }

  /**
   * Process revenue from a stream
   */
  async processRevenue(
    stream: string,
    amount: string,
    currency: string = '$EXS'
  ): Promise<any> {
    try {
      const response = await axios.post(`${this.apiEndpoint}/revenue/process`, {
        stream,
        amount,
        currency,
      });
      return response.data;
    } catch (error) {
      throw new Error(`Failed to process revenue: ${error}`);
    }
  }

  /**
   * Get all available revenue streams
   */
  getRevenueStreams(): RevenueStream[] {
    return [
      {
        name: 'cross_chain_mining',
        description: 'Mining across BTC, ETH, LTC, XMR, DOGE',
        treasuryShare: 0.4,
        userShare: 0.55,
        operationalShare: 0.05,
        estimatedApr: '8-15%',
        status: 'active',
      },
      {
        name: 'futures_trading',
        description: 'Automated futures on GMX, dYdX, Synthetix',
        treasuryShare: 0.3,
        userShare: 0.6,
        operationalShare: 0.1,
        estimatedApr: '12-25%',
        status: 'active',
      },
      {
        name: 'lightning_routing',
        description: 'P2TR Lightning channel routing fees',
        treasuryShare: 0.35,
        userShare: 0.6,
        operationalShare: 0.05,
        estimatedApr: '10-20%',
        status: 'active',
      },
      {
        name: 'taproot_processing',
        description: 'P2TR transaction batching and optimization',
        treasuryShare: 0.25,
        userShare: 0.7,
        operationalShare: 0.05,
        estimatedApr: '5-12%',
        status: 'active',
      },
      {
        name: 'yield_farming',
        description: 'Aave, Compound, Curve, Convex yield strategies',
        treasuryShare: 0.3,
        userShare: 0.65,
        operationalShare: 0.05,
        estimatedApr: '6-18%',
        status: 'active',
      },
      {
        name: 'mev_extraction',
        description: 'Flashbots and MEV-boost strategies',
        treasuryShare: 0.4,
        userShare: 0.5,
        operationalShare: 0.1,
        estimatedApr: '15-40%',
        status: 'active',
      },
      {
        name: 'staking_services',
        description: 'ETH, ADA, DOT, ATOM, SOL staking pools',
        treasuryShare: 0.2,
        userShare: 0.75,
        operationalShare: 0.05,
        estimatedApr: '4-12%',
        status: 'active',
      },
      {
        name: 'nft_royalties',
        description: 'Curated NFT collections with royalty sharing',
        treasuryShare: 0.3,
        userShare: 0.6,
        operationalShare: 0.1,
        estimatedApr: '8-25%',
        status: 'active',
      },
      {
        name: 'lending_protocol',
        description: 'Over-collateralized lending with BTC/ETH/USDC',
        treasuryShare: 0.25,
        userShare: 0.7,
        operationalShare: 0.05,
        estimatedApr: '5-15%',
        status: 'active',
      },
    ];
  }

  /**
   * Stub revenue stats for offline mode
   */
  private getStubRevenueStats(): RevenueStats {
    const streams = this.getRevenueStreams();
    const streamMap: Record<string, RevenueStream> = {};
    
    streams.forEach(stream => {
      streamMap[stream.name] = stream;
    });

    return {
      totalRevenueGenerated: '0',
      totalTreasuryCollected: '0',
      totalUserRewards: '0',
      activeStreams: streams.filter(s => s.status === 'active').length,
      totalStreams: streams.length,
      streams: streamMap,
    };
  }

  /**
   * Get income summary for user
   */
  async getIncomeSummary(address: string): Promise<any> {
    // This would query the blockchain/database in production
    return {
      address,
      totalEarned: '0.00000000',
      miningRewards: '0.00000000',
      stakingRewards: '0.00000000',
      revenueShare: '0.00000000',
      lastUpdate: new Date().toISOString(),
    };
  }

  /**
   * Estimate potential income based on parameters
   */
  estimatePotentialIncome(
    stake: number,
    miningHashRate: number,
    daysActive: number
  ): any {
    // Block reward: 50 EXS
    const blockReward = 50;
    const blocksPerDay = 144; // ~10 min blocks
    
    // Mining income (simplified)
    const dailyMiningReward = (miningHashRate / 1000) * blockReward * 0.01;
    
    // Staking income (simplified APY)
    const stakingApy = 0.08; // 8%
    const dailyStakingReward = (stake * stakingApy) / 365;
    
    // Revenue share (based on streams)
    const avgRevenueApy = 0.12; // 12%
    const dailyRevenueShare = (stake * avgRevenueApy) / 365;
    
    const totalDailyIncome = dailyMiningReward + dailyStakingReward + dailyRevenueShare;
    const projectedIncome = totalDailyIncome * daysActive;
    
    return {
      daily: {
        mining: dailyMiningReward.toFixed(8),
        staking: dailyStakingReward.toFixed(8),
        revenueShare: dailyRevenueShare.toFixed(8),
        total: totalDailyIncome.toFixed(8),
      },
      projected: {
        income: projectedIncome.toFixed(8),
        days: daysActive,
      },
      parameters: {
        stake,
        miningHashRate,
        stakingApy: `${(stakingApy * 100).toFixed(1)}%`,
        revenueApy: `${(avgRevenueApy * 100).toFixed(1)}%`,
      },
    };
  }
}

File: cmd/exs-node/revenue.go
package main

import (
	"fmt"

	"github.com/spf13/cobra"
)

var revenueCmd = &cobra.Command{
	Use:   "revenue",
	Short: "Multi-stream revenue operations",
	Long: `Manage and monitor the 9 revenue streams that fund the treasury:
  
  1. Cross-chain mining (BTC, ETH, LTC, XMR, DOGE)
  2. Smart contract futures trading
  3. Lightning fee routing
  4. Taproot transaction batching
  5. DeFi yield farming (Aave, Compound, Curve, Convex)
  6. MEV extraction (Flashbots, MEV-boost)
  7. Multi-chain staking (ETH, ADA, DOT, ATOM, SOL)
  8. NFT royalty pools
  9. $EXS lending protocol`,
}

var revenueShowCmd = &cobra.Command{
	Use:   "show",
	Short: "Show all revenue streams",
	Run: func(cmd *cobra.Command, args []string) {
		fmt.Println("💰 EXCALIBUR-EXS REVENUE STREAMS")
		fmt.Println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
		fmt.Println()
		
		streams := getRevenueStreams()
		
		for i, stream := range streams {
			fmt.Printf("%d. %s\n", i+1, stream.Name)
			fmt.Printf("   %s\n", stream.Description)
			fmt.Printf("   Status: %s | APR: %s\n", stream.Status, stream.EstimatedAPR)
			fmt.Printf("   Distribution: Treasury: %.0f%% | Users: %.0f%% | Ops: %.0f%%\n",
				stream.TreasuryShare*100,
				stream.UserShare*100,
				stream.OperationalShare*100,
			)
			fmt.Println()
		}
		
		fmt.Println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
		fmt.Println("Total Revenue Generated:    0.00 EXS")
		fmt.Println("Treasury Collected:         0.00 EXS")
		fmt.Println("User Rewards Distributed:   0.00 EXS")
		fmt.Println("✗ Live data not implemented yet")
	},
}

var revenueStatsCmd = &cobra.Command{
	Use:   "stats [stream-name]",
	Short: "Show statistics for a specific revenue stream",
	Args:  cobra.MaximumNArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		if len(args) == 0 {
			// Show summary stats
			fmt.Println("📊 Revenue Statistics Summary")
			fmt.Println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
			fmt.Println("Total Revenue (24h):     0.00 EXS")
			fmt.Println("Total Revenue (7d):      0.00 EXS")
			fmt.Println("Total Revenue (30d):     0.00 EXS")
			fmt.Println("Active Streams:          9")
			fmt.Println("Average APR:             10-20%")
			fmt.Println("✗ Not implemented yet")
			return
		}
		
		streamName := args[0]
		fmt.Printf("📊 Revenue Statistics: %s\n", streamName)
		fmt.Println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
		fmt.Println("Status:              Active")
		fmt.Println("Revenue (24h):       0.00 EXS")
		fmt.Println("Revenue (7d):        0.00 EXS")
		fmt.Println("Revenue (30d):       0.00 EXS")
		fmt.Println("Current APR:         0.00%")
		fmt.Println("Treasury Share:      0.00 EXS")
		fmt.Println("User Rewards:        0.00 EXS")
		fmt.Println("✗ Not implemented yet")
	},
}

var revenueEnableCmd = &cobra.Command{
	Use:   "enable [stream-name]",
	Short: "Enable a revenue stream",
	Args:  cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		streamName := args[0]
		fmt.Printf("Enabling revenue stream: %s\n", streamName)
		fmt.Println("✗ Not implemented yet")
	},
}

var revenueDisableCmd = &cobra.Command{
	Use:   "disable [stream-name]",
	Short: "Disable a revenue stream",
	Args:  cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		streamName := args[0]
		fmt.Printf("Disabling revenue stream: %s\n", streamName)
		fmt.Println("✗ Not implemented yet")
	},
}

var revenueDetailsCmd = &cobra.Command{
	Use:   "details [stream-name]",
	Short: "Show detailed information about a revenue stream",
	Args:  cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		streamName := args[0]
		
		stream := findRevenueStream(streamName)
		if stream == nil {
			fmt.Printf("Revenue stream not found: %s\n", streamName)
			return
		}
		
		fmt.Printf("💰 %s\n", stream.Name)
		fmt.Println("━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
		fmt.Printf("Description: %s\n", stream.Description)
		fmt.Printf("Status:      %s\n", stream.Status)
		fmt.Printf("Est. APR:    %s\n", stream.EstimatedAPR)
		fmt.Println("\nDistribution Model:")
		fmt.Printf("  Treasury:    %.0f%%\n", stream.TreasuryShare*100)
		fmt.Printf("  Users:       %.0f%%\n", stream.UserShare*100)
		fmt.Printf("  Operational: %.0f%%\n", stream.OperationalShare*100)
		
		fmt.Println("\nTechnical Details:")
		printStreamDetails(streamName)
	},
}

type RevenueStream struct {
	Name              string
	Description       string
	TreasuryShare     float64
	UserShare         float64
	OperationalShare  float64
	EstimatedAPR      string
	Status            string
}

func getRevenueStreams() []RevenueStream {
	return []RevenueStream{
		{
			Name:              "Cross-Chain Mining",
			Description:       "Mining across BTC, ETH, LTC, XMR, DOGE",
			TreasuryShare:     0.40,
			UserShare:         0.55,
			OperationalShare:  0.05,
			EstimatedAPR:      "8-15%",
			Status:            "🟢 Active",
		},
		{
			Name:              "Smart Contract Futures",
			Description:       "Automated futures on GMX, dYdX, Synthetix",
			TreasuryShare:     0.30,
			UserShare:         0.60,
			OperationalShare:  0.10,
			EstimatedAPR:      "12-25%",
			Status:            "🟢 Active",
		},
		{
			Name:              "Lightning Fee Routing",
			Description:       "P2TR Lightning channel routing fees",
			TreasuryShare:     0.35,
			UserShare:         0.60,
			OperationalShare:  0.05,
			EstimatedAPR:      "10-20%",
			Status:            "🟢 Active",
		},
		{
			Name:              "Taproot Processing",
			Description:       "P2TR transaction batching and optimization",
			TreasuryShare:     0.25,
			UserShare:         0.70,
			OperationalShare:  0.05,
			EstimatedAPR:      "5-12%",
			Status:            "🟢 Active",
		},
		{
			Name:              "DeFi Yield Farming",
			Description:       "Aave, Compound, Curve, Convex yield strategies",
			TreasuryShare:     0.30,
			UserShare:         0.65,
			OperationalShare:  0.05,
			EstimatedAPR:      "6-18%",
			Status:            "🟢 Active",
		},
		{
			Name:              "MEV Extraction",
			Description:       "Flashbots and MEV-boost strategies",
			TreasuryShare:     0.40,
			UserShare:         0.50,
			OperationalShare:  0.10,
			EstimatedAPR:      "15-40%",
			Status:            "🟢 Active",
		},
		{
			Name:              "Staking Services",
			Description:       "ETH, ADA, DOT, ATOM, SOL staking pools",
			TreasuryShare:     0.20,
			UserShare:         0.75,
			OperationalShare:  0.05,
			EstimatedAPR:      "4-12%",
			Status:            "🟢 Active",
		},
		{
			Name:              "NFT Royalty Pools",
			Description:       "Curated NFT collections with royalty sharing",
			TreasuryShare:     0.30,
			UserShare:         0.60,
			OperationalShare:  0.10,
			EstimatedAPR:      "8-25%",
			Status:            "🟢 Active",
		},
		{
			Name:              "$EXS Lending Protocol",
			Description:       "Over-collateralized lending with BTC/ETH/USDC",
			TreasuryShare:     0.25,
			UserShare:         0.70,
			OperationalShare:  0.05,
			EstimatedAPR:      "5-15%",
			Status:            "🟢 Active",
		},
	}
}

func findRevenueStream(name string) *RevenueStream {
	streams := getRevenueStreams()
	for i := range streams {
		if streams[i].Name == name {
			return &streams[i]
		}
	}
	return nil
}

func printStreamDetails(name string) {
	details := map[string]map[string]string{
		"Cross-Chain Mining": {
			"Chains":     "BTC, ETH, LTC, XMR, DOGE",
			"Algorithm":  "Multi-algo support",
			"Pool":       "Distributed across multiple pools",
		},
		"Lightning Fee Routing": {
			"Channels":   "P2TR Lightning channels",
			"Network":    "Lightning Network",
			"Fees":       "Dynamic routing fees",
		},
		"DeFi Yield Farming": {
			"Protocols":  "Aave, Compound, Curve, Convex",
			"Strategy":   "Automated yield optimization",
			"Tokens":     "USDC, DAI, WBTC, ETH",
		},
	}
	
	if d, ok := details[name]; ok {
		for k, v := range d {
			fmt.Printf("  %s: %s\n", k, v)
		}
	} else {
		fmt.Println("  Details available after integration")
	}
}

func init() {
	revenueCmd.AddCommand(
		revenueShowCmd,
		revenueStatsCmd,
		revenueEnableCmd,
		revenueDisableCmd,
		revenueDetailsCmd,
	)
	
	rootCmd.AddCommand(revenueCmd)
}
